### PR TITLE
(#4) Windows processes snapshot

### LINKED ISSUE #4: Windows: more consistent API
Currently, we rely on WMI events to list Windows processes. That API is good enough for most purposes, but it has a problem with consistency: if we only rely on process start/stop events, we may easily miss some processes.

Also, we don't show the processes started before the program startup.

I suggest that we consider using periodic requests to `CreateToolhelp32Snapshot`, or periodic WMI queries of the full process list.

`CreateToolhelp32Snapshot` has a downside that we won't be able to get the process' command line.

The latter I remember as being clunky, but this is yet to be checked :)

Also, if we migrate to WMI list queries instead of WMI events, we may easily miss some processes, and we don't want to miss any (we should properly register quick start+exit of a process, if possible).

### PR DESCRIPTION
**Pull request by y0ung3r**
**Should close #4**

**Changed:**
- `ProcessModel` has been renamed to `SystemProcess` (because `SystemProcess` will be the extension point for each operating system)
- Created a `WindowsProcess` (`SystemProcess` inheritor) model that encapsulates factories for creating a process from `WinAPI` structures and `WMI` objects
- Implemented `ProcessesSnapshot` class to get Windows processes snapshot
- Added `PInvoke` dependency for interaction with `WinAPI` (`PInvoke` is no longer maintained. In future, we should use [CsWin32](https://github.com/microsoft/CsWin32) when stable version is released)

**Screenshot:**

image

### CODE REVIEW

File: ProcessDoctor.Backend.Windows/NT/Extensions/ProcessBasicInformationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+namespace ProcessDoctor.Backend.Windows.NT.Extensions;
+
+internal static class ProcessBasicInformationExtensions
+{
+    internal static unsafe
+#if TARGET_64BIT
+        PEB_64
+#else
+        PEB_32
+#endif
```
Comment (ForNeVeR) [inline ProcessDoctor.Backend.Windows/NT/Extensions/ProcessBasicInformationExtensions.cs]: _(Minor)_ Generally, I'd say we are not interested in the 32-bit Windows target. Is it even supported on modern Windows versions? I suggest to just get rid of it.

File: ProcessDoctor.Backend.Windows/NT/Extensions/ProcessHandleExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+using ProcessDoctor.Backend.Windows.Exceptions;
+
+namespace ProcessDoctor.Backend.Windows.NT.Extensions;
+
+internal static class ProcessHandleExtensions
+{
+    internal static NTDll.PROCESS_BASIC_INFORMATION GetBasicInformation(this Kernel32.SafeObjectHandle processHandle)
+    {
+        var length = Marshal.SizeOf<NTDll.PROCESS_BASIC_INFORMATION>();
+        var pointer = Marshal.AllocHGlobal(length);
```
Comment (ForNeVeR) [inline ProcessDoctor.Backend.Windows/NT/Extensions/ProcessHandleExtensions.cs]: Here you don't need to allocate the memory using `Marshal.AllocHGlobal`. You could just create a structure on the stack and pass its address to `NtQueryInformationProcess`, right?

File: ProcessDoctor.Backend.Windows/NT/Extensions/UnicodeStringExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        if (unmanagedString.Length == 0 || unmanagedString.MaximumLength == 0)
+        {
+            return null;
+        }
+
+        var bytes = memoryReader.ReadBytes(
+            unmanagedString.Buffer_IntPtr,
+            unmanagedString.Length);
+
+        return Encoding.Unicode.GetString(bytes);
```
Comment (ForNeVeR) [inline ProcessDoctor.Backend.Windows/NT/Extensions/UnicodeStringExtensions.cs]: Let's try something like this instead:
```
if (unmanagedString.Length % 2 != 0) {
  // malformed UTF-16 string
  throw new CorruptedProcessHandleException();
}

var span = new Span<char>(unmanagedString.Buffer, unmanagedString.Length / 2);
return new string(span);
```

This should be more efficient.

File: ProcessDoctor.Backend.Windows/ProcessMemoryReader.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        var isSuccess = Kernel32.ReadProcessMemory(
+            _processHandle,
+            address,
+            buffer,
+            new UIntPtr((uint)length),
+            out var totalLength);
+
+        return isSuccess && length == totalLength.ToUInt32();
+    }
```
Comment (ForNeVeR) [inline ProcessDoctor.Backend.Windows/ProcessMemoryReader.cs]: Here I'd prefer the following implementation:
```csharp
private unsafe bool TryRead(IntPtr address, Span<byte> resultArray)
{
    fixed (byte* data = resultArray)
    {
        var isSuccess = Kernel32.ReadProcessMemory(
            _processHandle,
            address,
            (IntPtr)data,
            (nuint)resultArray.Length,
            out var totalLength);

        return isSuccess && (nuint)resultArray.Length == totalLength;
    }
}
```

Note how it accepts a managed array and thus no longer requires `AllocHGlobal` and all that.

You'd be able to call it like this:
```csharp
public T? ReadStructure<T>(IntPtr address)
    where T : unmanaged
{
    T instance = default;
    if (!TryRead(address, MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref instance, 1))))
    {
        return null;
    }

    return instance;
}
```

(I didn't test this approach, but I think it should work, heh)

**Review by ForNeVeR — CHANGES_REQUESTED** [review]
Thanks! This is great. I have some small concerns, please address them, and this will be ready for merge.

File: ProcessDoctor.Backend.Windows/NT/RTL_USER_PROCESS_PARAMETERS.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+using PInvoke;
+
+namespace ProcessDoctor.Backend.Windows.NT;
+
+[StructLayout(LayoutKind.Sequential)]
+public unsafe struct RTL_USER_PROCESS_PARAMETERS
+{
+    private fixed byte Reserved1[16];
+
+    private fixed int Reserved2[20];
```
Comment (y0ung3r) [inline ProcessDoctor.Backend.Windows/NT/RTL_USER_PROCESS_PARAMETERS.cs]: Honestly, I don't know how to write a fixed `PVOID` array any other way. Maybe you have some ideas?

**RTL_USER_PROCESS_PARAMETERS structure:**
```c
typedef struct _RTL_USER_PROCESS_PARAMETERS {
  BYTE           Reserved1[16];
  PVOID          Reserved2[10];
  UNICODE_STRING ImagePathName;
  UNICODE_STRING CommandLine;
} RTL_USER_PROCESS_PARAMETERS, *PRTL_USER_PROCESS_PARAMETERS;
```

File: ProcessDoctor.Backend.Windows/NT/Extensions/ProcessHandleExtensions.cs (hunk 2 above)
Comment (y0ung3r) [reply ProcessDoctor.Backend.Windows/NT/Extensions/ProcessHandleExtensions.cs]: That's right

File: ProcessDoctor.Backend.Windows/ProcessMemoryReader.cs (hunk 4 above)
Comment (y0ung3r) [reply ProcessDoctor.Backend.Windows/ProcessMemoryReader.cs]: I got rid of `ProcessMemoryReader` and put all that code in the extension methods

File: ProcessDoctor.Backend.Windows/NT/RTL_USER_PROCESS_PARAMETERS.cs (hunk 5 above)
Comment (y0ung3r) [reply ProcessDoctor.Backend.Windows/NT/RTL_USER_PROCESS_PARAMETERS.cs]: I replaced `int` with `long` and returned the standard array size:

```csharp
private fixed long Reserved2[10];
```

It seems to be working 😁

**Review by ForNeVeR — APPROVED** [review]
Thanks for contribution!

**Pull request merged**